File: algorithms/python/300.longest-increasing-subsequence.py
#
# @lc app=leetcode id=300 lang=python3
#
# [300] Longest Increasing Subsequence
#
# Given an integer array nums, return the length of the longest strictly
# increasing subsequence.
#
# Constraints:
#   • 1 <= nums.length <= 2500
#   • -10^4 <= nums[i] <= 10^4
#

# @lc code=start

class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:

        # Complexity:
        #   • Time: O(n^2)
        #   • Space: O(n)

        n = len(nums)
        dp = [1] * n

        for i in range(1, len(nums)):
            for j in range(0, i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[j] + 1, dp[i])

        return max(dp)

# @lc code=end

File: algorithms/cpp/300.longest-increasing-subsequence.cpp
/*
 * @lc app=leetcode id=300 lang=cpp
 *
 * [300] Longest Increasing Subsequence
 *
 * Given an integer array nums, return the length of the longest strictly
 * increasing subsequence.
 *
 * Constraints:
 *   • 1 <= nums.length <= 2500
 *   • -10^4 <= nums[i] <= 10^4
 */

// @lc code=start

#include <algorithm>
#include <vector>

class Solution {
 public:
  int lengthOfLIS(const std::vector<int> &nums) {

    // Complexity:
    //   • Time: O(n^2)
    //   • Space: O(n)

    size_t n = nums.size();
    std::vector<int> dp(n, 1);

    for (size_t i = 1; i < n; ++i) {
      for (size_t j = 0; j < i; ++j) {
        if (nums[i] > nums[j]) {
          dp[i] = std::max(dp[j] + 1, dp[i]);
        }
      }
    }

    return *std::max_element(dp.begin(), dp.end());
  }
};

// @lc code=end
